client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/query/RemoteQuery.java

```diff
@@ -88,19 +93,20 @@ public CloseableIterator<T> iterator() {
       if (maxResults == -1 && startOffset == 0) {
          log.warnPerfRemoteIterationWithoutPagination(queryString);
       }
-      return Closeables.iterator(execute().list().iterator());
+      return Closeables.iterator(list().iterator());
    }
 
    @Override
    public int getResultSize() {
-      BaseQueryResponse<?> response = executeRemotely();
+      BaseQueryResponse<?> response = executeRemotely(true);
       return (int) response.getTotalResults();
    }
 
-   private BaseQueryResponse<T> executeRemotely() {
+   private BaseQueryResponse<T> executeRemotely(boolean withHitCount) {
       validateNamedParameters();
-      QueryOperation op = cache.getOperationsFactory().newQueryOperation(this, cache.getDataFormat());
-      return (BaseQueryResponse<T>) (timeout != -1 ? await(op.execute(), TimeUnit.NANOSECONDS.toMillis(timeout)) : await(op.execute()));
+      QueryOperation op = cache.getOperationsFactory().newQueryOperation(this, cache.getDataFormat(), withHitCount);
+      return (BaseQueryResponse<T>) (timeout != -1 ? await(op.execute(), TimeUnit.NANOSECONDS.toMillis(timeout)) :
```

Comment:
Is this invoked in a non blocking thread or is it always a blocking thread? From the look of it, it seems like possibly a non blocking thread.

Reply:
I think that we're in the client VM.
Should our HotRod client APIs be always non blocking for all methods we expose?
In case we need to change many of these methods.
In any case I don't think that this contribution has introduced the blocking call, so I would handle that in a different PR.

Reply:
Technically yes it does and prevents doing any operations to a server that is tied to that thread. We should definitely remedy this in another PR.